keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1065,6 +1065,51 @@ Details to consider:
 
 #### CEL Expression Composition
 
+##### Policy Predicates
+
+The match criteria in bindings are not expected to be able to cover all possible
+ways users may want to scope their policies. For example, there is no way to
+match off of kind, only resource. To provide extensibility for the match critiera
+without requiring modifying every validation rule individually, a global predicate
+system is needed. These predicates contain CEL statements that must be satisfied, otherwise
+the policy will be ignored. In order to keep bindings language-agnostic and to support
+singleton policies, the logic should live in the policy definition resource. To enable
+customization per-binding, the CEL statements should have access to the parameter resource.
+
+Here is an example of a policy definition using policy predicates (under the `customConstraints` field):
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "replicalimit-policy.example.com"
+Spec:
+  failurePolicy: Fail
+  paramKind:
+    apiVersion: rules.example.com/v1
+    kind: ReplicaLimit
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+  customConstraints:
+    - expression: 'metadata.kind == "Deployment"'
+    - expression: '!(metadata.namespace in params.excludedNamespaces)'
+  validations:
+    - expression: "object.spec.replicas <= params.maxReplicas"
+      reason: Invalid
+```
+
+For demonstration purposes, we assume `match` has no support for `excludedNamespaces`.
+
+`customConstraints` has the following behaviors:
+
+* Only the request object and parameters are accessible (no referential lookup)
```

Comment:
I don't know if this belongs in the KEP but we can statically determine of a CEL expression contains the `params` variable. If it does, we would evaluate the expression for each binding. If it does not, we could validate it once for all bindings. This would be an optimization.

Reply:
+1, maybe call it out as a possible future optimization in a code comment once we get a sense of how much real-world cost it'd save?